### ISSUE TITLE
Can benchmarking functionality have used externalities extensions? 

### ISSUE THREAD

**Issue by kianenigma** [paritytech/substrate#12591]
Having a talk with @cheme about externalities, it appears that a lot of the optional host functions that can be used in some context and not others are declared as: 

```
#[cfg(feature = "std")]
sp_externalities::decl_extension! {
	/// An externalities extension to submit transactions to the pool.
	pub struct TransactionPoolExt(Box<dyn TransactionPool + Send>);
}
```

But it seems like the benchmarking code is declared right into the main `trait Externalities`.

Why is this? could it have been done otherwise?

**Comment by bkchr** [comment]
No. It currently could not. The problem is that you need access to the overlay and that overlay is already part of the object that implements the externalities crate. However, if we move everything over to just use extensions, we probably could do the switch for benchmarking to have their own extension and I would like this.

**Comment by ggwpez** [comment]
You are talking specifically about these functions here?  
https://github.com/paritytech/substrate/blob/c8653447fc8ef8d95a92fe164c96dffb37919e85/primitives/externalities/src/lib.rs#L282-L287  

We could try feature gating them at least. But that would introduce the `runtime-benchmarks` feature into some crates which normally dont have it, like `sp-state-machine`.  
Do you think its worth it?

**Comment by bkchr** [comment]
I don't think it is wort it. I would like to solve it as I pointed it out above.
